### PR TITLE
Ports panic bunker function from Hestia

### PR DESCRIPTION
**Pull request by quardbreak**
Adds the panic bunker from Hestia. It lack of SSbabysit which I personally don't like, so I ported only PB since it will be useful for hi-pop servers.

## Changelog
:cl:
rscadd: Added the panic bunker support from Hestia repository. By default, it turned off and can be toggled on in config or manually for a round.
admin: Admins can control the panic bunker state and add/revoke bypass in Server section with R_SERVER flag.
/:cl:

### CODE REVIEW

File: code/modules/admin/panicbunker.dm [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+	log_admin("[key_name(usr)] has added [ckeytobypass] to the current round's bunker bypass list.")
+	message_admins("[key_name_admin(usr)] has added [ckeytobypass] to the current round's bunker bypass list.")
+
+/datum/admins/proc/revokebunkerbypass(ckeytobypass as text)
+	set category = "Server"
+	set name = "Revoke Panic Bunker Bypass"
+	set desc = "Revoke's a ckey's permission to bypass the panic bunker for a given round."
+	if(!dbcon.IsConnected())
+		to_chat(usr, "<span class='adminnotice'>The Database is not enabled or not working!</span>")
+		return
```
Comment (afterthought2) [inline code/modules/admin/panicbunker.dm]: Why make this check when we don't actually store this to the db?

File: code/modules/client/client_procs.dm [hunk 2]
```diff
@@ -325,6 +325,14 @@
...
 	var/sql_admin_rank = sql_sanitize_text(admin_rank)
 
+	if ((player_age == -1) && !(ckey in global.panic_bunker_bypass)) //first connection
+		if (config.panic_bunker && !holder && !deadmin_holder)
+			log_adminwarn("Failed Login: [key] - New account attempting to connect during panic bunker")
+			message_admins("<span class='adminnotice'>Failed Login: [key] - New account attempting to connect during panic bunker</span>")
+			to_chat(src, config.panic_bunker_message)
+			qdel(src)
+			return 0
+		player_age = 0 //math requires this to not be -1.
```
Comment (afterthought2) [inline code/modules/client/client_procs.dm]: Why are we kicking clients in the middle of a logging helper?

File: code/modules/client/client_procs.dm [hunk 3]
```diff
@@ -325,6 +325,14 @@
 	var/sql_computerid = sql_sanitize_text(src.computer_id)
 	var/sql_admin_rank = sql_sanitize_text(admin_rank)
 
+	if ((player_age == -1) && !(ckey in global.panic_bunker_bypass)) //first connection
```
Comment (afterthought2) [inline code/modules/client/client_procs.dm]: It seems this branch will never be traversed, as `player_age` is set to `0` on line 285 (unless the db returns an explicit `-1`, which also seems impossible).

File: config/example/config.txt [hunk 4]
```diff
@@ -426,6 +426,9 @@ RADIATION_LOWER_LIMIT 0.15
 ## Uncomment this to modify the number of actions permitted per interval before being kicked for spam.
 #MAX_ACTS_PER_INTERVAL 140
 
+## Is the panic bunker currently on?
+#PANIC_BUNKER
```
Comment (afterthought2) [inline config/example/config.txt:430]: Should be annotated as "on by default" or similar; it's a config setting, not a state var.

File: code/modules/admin/panicbunker.dm [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+	set category = "Server"
+	set name = "Add Panic Bunker Bypass"
+	set desc = "Allows a given ckey to connect despite the panic bunker for a given round."
+
+	if(!check_rights(R_SERVER))
+		return
+
+	if(!dbcon.IsConnected())
+		to_chat(usr, "<span class='adminnotice'>The Database is not enabled or not working!</span>")
+		return
```
Comment (afterthought2) [inline code/modules/admin/panicbunker.dm]: Again, why?

File: code/controllers/configuration.dm [hunk 6]
```diff
@@ -236,6 +236,9 @@ var/list/gamemode_cache = list()
 	var/max_acts_per_interval = 140 //Number of actions per interval permitted for spam protection.
 	var/act_interval = 0.1 SECONDS //Interval for spam prevention.
 
+	var/panic_bunker = FALSE //is the panic bunker enabled?
+	var/panic_bunker_message = "Sorry! The panic bunker is enabled. Please head to our discord or forum to get yourself added to the panic bunker bypass."
```
Comment (afterthought2) [inline code/controllers/configuration.dm]: Why is this a var on the config controller, but cannot actually be set in the config?

**Review by afterthought2 — COMMENTED** [review]
Not a big fan of the lack of automation here.

File: code/controllers/configuration.dm (hunk 6 above)
Comment (quardbreak) [reply code/controllers/configuration.dm]: oversight, will make it configruable later.

Comment (MistakeNot4892) [reply code/controllers/configuration.dm]: `Discord` needs a capital.

File: code/modules/admin/panicbunker.dm [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+	if(!check_rights(R_SERVER))
+		return
+
+	global.panic_bunker_bypass |= ckey(ckeytobypass)
+
+	log_admin("[key_name(usr)] has added [ckeytobypass] to the current round's bunker bypass list.")
+	message_admins("[key_name_admin(usr)] has added [ckeytobypass] to the current round's bunker bypass list.")
+
+/datum/admins/proc/revokebunkerbypass(ckeytobypass as text)
```
Comment (MistakeNot4892) [inline code/modules/admin/panicbunker.dm]: `ckeytobypass in global.panic_bunker_bypass` would let them just pick one to remove I believe.

File: code/modules/admin/panicbunker.dm [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+	set category = "Server"
+	set name = "Toggle Panic Bunker"
+
+	if(!check_rights(R_SERVER))
+		return
+
+	config.panic_bunker = (!config.panic_bunker)
+
+	log_and_message_admins("[key_name(usr)] has toggled the Panic Bunker, it is now [(config.panic_bunker ? "on" : "off")]")
+	if (config.panic_bunker && (!dbcon || !dbcon.IsConnected()))
```
Comment (MistakeNot4892) [inline code/modules/admin/panicbunker.dm:13]: Checking for db here again.

File: code/modules/admin/panicbunker.dm [hunk 9]
```diff
@@ -0,0 +1,37 @@
+var/list/panic_bunker_bypass = list()
+
+/datum/admins/proc/panicbunker()
+	set category = "Server"
+	set name = "Toggle Panic Bunker"
+
+	if(!check_rights(R_SERVER))
+		return
+
+	config.panic_bunker = (!config.panic_bunker)
```
Comment (MistakeNot4892) [inline code/modules/admin/panicbunker.dm]: Why `()`?

File: code/modules/admin/panicbunker.dm (hunk 8 above)
Comment (MistakeNot4892) [reply code/modules/admin/panicbunker.dm:13]: Admin note got updated after discussion on Discord, my brain is smooth

**Review by MistakeNot4892 — APPROVED** [review]

File: code/modules/admin/panicbunker.dm [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+/datum/admins/proc/panicbunker()
+	set category = "Server"
+	set name = "Toggle Panic Bunker"
+
+	if(!check_rights(R_SERVER))
+		return
+
+	config.panic_bunker = !config.panic_bunker
+
+	log_and_message_admins("[key_name(usr)] has toggled the Panic Bunker, it is now [(config.panic_bunker ? "on" : "off")]")
```
Comment (PsiOmegaDelta) [inline code/modules/admin/panicbunker.dm:12]: `log_and_message_admins("has toggled the Panic Bunker, it is now [(config.panic_bunker ? "on" : "off")]"`
`usr` is automatically prepended by `log_and_message_admins()`

File: code/modules/admin/panicbunker.dm [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+	set name = "Add Panic Bunker Bypass"
+	set desc = "Allows a given ckey to connect despite the panic bunker for a given round."
+
+	if(!check_rights(R_SERVER))
+		return
+
+	global.panic_bunker_bypass |= ckey(ckeytobypass)
+
+	log_admin("[key_name(usr)] has added [ckeytobypass] to the current round's bunker bypass list.")
+	message_admins("[key_name_admin(usr)] has added [ckeytobypass] to the current round's bunker bypass list.")
```
Comment (PsiOmegaDelta) [inline code/modules/admin/panicbunker.dm:27]: Replace both log lines above with
`log_and_message_admins("has added [ckeytobypass] to the current round's bunker bypass list.")`

File: code/modules/admin/panicbunker.dm [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+
+/datum/admins/proc/revokebunkerbypass(ckeytobypass in global.panic_bunker_bypass)
+	set category = "Server"
+	set name = "Revoke Panic Bunker Bypass"
+	set desc = "Revoke's a ckey's permission to bypass the panic bunker for a given round."
+
+	global.panic_bunker_bypass -= ckey(ckeytobypass)
+
+	log_admin("[key_name(usr)] has removed [ckeytobypass] from the current round's bunker bypass list.")
+	message_admins("[key_name_admin(usr)] has removed [ckeytobypass] from the current round's bunker bypass list.")
```
Comment (PsiOmegaDelta) [inline code/modules/admin/panicbunker.dm:37]: As above, replace both lines with
`log_and_message_admins("has removed [ckeytobypass] from the current round's bunker bypass list.")`

**Pull request merged**